python/ray/dag/dag_node.py

```diff
@@ -335,6 +354,33 @@ def __call__(self, node: "DAGNode"):
             )
         )
 
+    def bfs(self, fn: "Callable[[DAGNode], T]"):
+        """Breadth-first search on the DAG."""
+        visited = set()
+        queue = [self]
+        while queue:
+            node = queue.pop(0)
+            if node not in visited:
+                fn(node)
+                visited.add(node)
+                # Add all unseen downstream and upstream nodes to the queue.
+                # This function should be called by the root of the DAG. However,
+                # in some invalid cases, some nodes may not be descendants of the
+                # root. Therefore, we also add upstream nodes to the queue so that
+                # a meaningful error message can be raised when the DAG is compiled.
+                for node in node._downstream_nodes + node._upstream_nodes:
+                    if node not in visited:
+                        queue.append(node)
+
+    def _find_root(self) -> "DAGNode":
+        """Return the root node of the DAG."""
+        root = self
+        while type(root).__name__ != "InputNode":
```

Comment:
why not just using isinstance?

Reply:
I copied from https://github.com/ray-project/ray/blob/master/python/ray/dag/dag_node.py#L318